### PR TITLE
[GLUTEN-842][VL] convert expand op to expand exec in velox

### PR DESCRIPTION
**Pull request by zhli1142015**
## What changes were proposed in this pull request?

Here are project expressions we observed in expand operation:

agg exprs + group by exprs + gid.
agg exprs + group by exprs + gid + _gen_grouping_pos. --> The last column is for handling duplicate grouping sets.
group by exprs + gid + agg exprs. --> gid is calculated by different way from above two cases. it's assigned with the sequence number of project set.

Original ExpandExecTrandofrmer can only handle the first case. I'm adding the expand exec in velox side by the PR: https://github.com/oap-project/velox/pull/199/files. This PR is for converting spark ExpandExec to Expand OP in Velox, We don't need to do the columns' mapping in Gluten.

The original ExpandExecTrandofrmer is renamed to GroupIdExecTrandofrmer to not break ClicKHouse.

## How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by JkSelf** [comment]
@baibaichen 

This PR is to implement ExpandTransformer similar with vanilla spark. And then we no need to distinguish the agg cols and group cols in Gluten. It seems the Expand [PR](https://github.com/substrait-io/substrait/pull/368) to substrait community is ready to merge except the doc update. How about directly use  [PR](https://github.com/substrait-io/substrait/pull/368) in Gluten to support Expand  in this PR?

**Comment by FelixYBW** [comment]
> @baibaichen
> 
> This PR is to implement ExpandTransformer similar with vanilla spark. And then we no need to distinguish the agg cols and group cols in Gluten. It seems the Expand [PR](https://github.com/substrait-io/substrait/pull/368) to substrait community is ready to merge except the doc update. How about directly use [PR](https://github.com/substrait-io/substrait/pull/368) in Gluten to support Expand in this PR?


We should follow substrait's solution. Does the PR need change to substrait?

**Comment by JkSelf** [comment]
> > @baibaichen
> > This PR is to implement ExpandTransformer similar with vanilla spark. And then we no need to distinguish the agg cols and group cols in Gluten. It seems the Expand [PR](https://github.com/substrait-io/substrait/pull/368) to substrait community is ready to merge except the doc update. How about directly use [PR](https://github.com/substrait-io/substrait/pull/368) in Gluten to support Expand in this PR?
> 
> We should follow substrait's solution. Does the PR need change to substrait?

Yes. This PR add new ExpandRel message in algebra.proto, which follow up the substrait community solution except the definition in [PR](https://github.com/substrait-io/substrait/pull/368).

**Comment by zhli1142015** [comment]
> > > @baibaichen
> > > This PR is to implement ExpandTransformer similar with vanilla spark. And then we no need to distinguish the agg cols and group cols in Gluten. It seems the Expand [PR](https://github.com/substrait-io/substrait/pull/368) to substrait community is ready to merge except the doc update. How about directly use [PR](https://github.com/substrait-io/substrait/pull/368) in Gluten to support Expand in this PR?
> > 
> > 
> > We should follow substrait's solution. Does the PR need change to substrait?
> 
> Yes. This PR add new ExpandRel message in algebra.proto, which follow up the substrait community solution except the definition in [PR](https://github.com/substrait-io/substrait/pull/368).

Thanks, updated.

**Comment by zhli1142015** [comment]
@JkSelf , @zhouyuan  and @lgbo-ustc, could you please help to take a look?

File: gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala [hunk 1]
```diff
@@ -84,160 +81,148 @@ case class ExpandExecTransformer(projections: Seq[Seq[Expression]],
...
-        .replaceWithExpressionTransformer(
-          projections.head(i),
-          originalInputAttributes)
-        .doTransform(args)
-      aggExprNodes.add(aggExprNode)
+    def needsPreProjection(projections: Seq[Seq[Expression]]): Boolean = {
+      projections
+        .exists(set => set.exists(p => !p.isInstanceOf[Attribute] && !p.isInstanceOf[Literal]))
     }
+    if (needsPreProjection(projections)) {
```
Comment (JkSelf) [inline gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala:93]: Can we add simple unit test for this case?

**Review by JkSelf — DISMISSED** [review]
LGTM except small comments.

**Review by FelixYBW — APPROVED** [review]

File: gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala (hunk 1 above)
Comment (zhli1142015) [reply gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala:93]: Added, thanks.
image

**Comment by lgbo-ustc** [comment]
It seems the `ExpandRel` has been changed, is it backward compatible with `ClickhHouse`?

**Comment by zhli1142015** [comment]
> It seems the `ExpandRel` has been changed, is it backward compatible with `ClickhHouse`?

new 'ExpandRel' is different and can't be compatible with CH. The original Rel is renamed to 'GroupIdRel' and changes are made in CH also.
Thanks.

File: gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto [hunk 2]
```diff
@@ -355,9 +355,38 @@ message ExchangeRel {
...
   Rel input = 2;
+  repeated ExpandField fields = 4;
+  substrait.extensions.AdvancedExtension advanced_extension = 10;
+
+  message ExpandField {
+    oneof field_type {
+      // Field that switches output based on which duplicate_id we're outputting
+      SwitchingField switching_field = 2;
+
+      // Field that outputs the same value no matter which duplicate_id we're on.
```
Comment (lgbo-ustc) [inline gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto:373]: Is there case that this will be used?

**Review by lgbo-ustc — COMMENTED** [review]
updated

**Comment by zhli1142015** [comment]
> It's greate. This implementation is simple. I think it could solve some problems we have meet.
> 
> LGTM.

Thanks for review @lgbo-ustc , will you work on CH side to consume new ExpandRel contract?

**Comment by lgbo-ustc** [comment]
> > It's greate. This implementation is simple. I think it could solve some problems we have meet.
> > LGTM.
> 
> Thanks for review @lgbo-ustc , will you work on CH side to consume new ExpandRel contract?

We will do it soon

**Review by zhouyuan — APPROVED** [review]
👍

**Comment by zhanglistar** [comment]
@zhli1142015 

```
agg exprs + group by exprs + gid + _gen_grouping_pos. --> The last column is for handling duplicate grouping sets.
group by exprs + gid + agg exprs. --> gid is calculated by different way from above two cases. it's assigned with the sequence number of project set.
```
I have a question, the two cases, could you give some sql? We don't know when spark will generate the two cases. 
Thanks.

**Comment by zhli1142015** [comment]
> @zhli1142015
> 
> ```
> agg exprs + group by exprs + gid + _gen_grouping_pos. --> The last column is for handling duplicate grouping sets.
> group by exprs + gid + agg exprs. --> gid is calculated by different way from above two cases. it's assigned with the sequence number of project set.
> ```
> 
> I have a question, the two cases, could you give some sql? We don't know when spark will generate the two cases. Thanks.

Hello @zhanglistar ,
Please check below sample code:
```
case class TestData3(a: Int, b: Option[Int])
val df = spark.sparkContext.parallelize(
      TestData3(1, None) ::
      TestData3(2, Some(2)) :: Nil).toDF()
import org.apache.spark.sql.functions._
df.agg(count($"a"), count($"b"), count(lit(1)), count_distinct($"a"), count_distinct($"b")).collect // case 3
df.createOrReplaceTempView("df")
spark.sql("select count(a) from df group by grouping sets((a), (a), (b))").collect // case 2
```

**Comment by zhanglistar** [comment]
> > @zhli1142015
> > ```
> > agg exprs + group by exprs + gid + _gen_grouping_pos. --> The last column is for handling duplicate grouping sets.
> > group by exprs + gid + agg exprs. --> gid is calculated by different way from above two cases. it's assigned with the sequence number of project set.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I have a question, the two cases, could you give some sql? We don't know when spark will generate the two cases. Thanks.
> 
> Hello @zhanglistar , Please check below sample code:
> 
> ```
> case class TestData3(a: Int, b: Option[Int])
> val df = spark.sparkContext.parallelize(
>       TestData3(1, None) ::
>       TestData3(2, Some(2)) :: Nil).toDF()
> import org.apache.spark.sql.functions._
> df.agg(count($"a"), count($"b"), count(lit(1)), count_distinct($"a"), count_distinct($"b")).collect // case 3
> df.createOrReplaceTempView("df")
> spark.sql("select count(a) from df group by grouping sets((a), (a), (b))").collect // case 2
> ```

@zhli1142015 Thanks! For the sql `spark.sql("select count(a) from df group by grouping sets((a), (a), (b))").collect // case 2`, why not just duplicate the two grouping sets (a) for optimization?

**Pull request merged**